### PR TITLE
issue2067: Follow gRPC error handing in associating gRPC error codes with errors

### PR DESCRIPTION
**Pull request by beiwei30**
# Description

fix issue #2067

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2067

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/grpc/api.go [hunk 1]
```diff
@@ -180,21 +179,21 @@ func (a *api) CallActor(ctx context.Context, in *internalv1pb.InternalInvokeRequ
 
 func (a *api) PublishEvent(ctx context.Context, in *runtimev1pb.PublishEventRequest) (*empty.Empty, error) {
 	if a.publishFn == nil {
-		err := errors.New("ERR_PUBSUB_NOT_FOUND")
+		err := status.Error(codes.Unavailable, "pubsub function is not configured")
```
Comment (youngbupark) [inline pkg/grpc/api.go:182]: Please follow this spec - https://cloud.google.com/apis/design/errors#handling_errors

I think `FAILED_PRECONDITION` makes more sense than Unavaiable. 

FAILED_PRECONDITION, "no pubsub is configured"

File: pkg/grpc/api.go [hunk 2]
```diff
@@ -327,11 +326,11 @@ func (a *api) GetBulkState(ctx context.Context, in *runtimev1pb.GetBulkStateRequ
 
 func (a *api) getStateStore(name string) (state.Store, error) {
 	if a.stateStores == nil || len(a.stateStores) == 0 {
-		return nil, errors.New("ERR_STATE_STORE_NOT_CONFIGURED")
+		return nil, status.Error(codes.Unavailable, "state store is not configured")
```
Comment (youngbupark) [inline pkg/grpc/api.go:329]: same here - FAILED_PRECONDITION

**Review by youngbupark — CHANGES_REQUESTED** [review]
Thanks for your contribution! I leave some comments.

**Comment by youngbupark** [comment]
@mukundansundar @ryanvolum @vinayada1 Please review this PR. This is the PR to correct non-standard gRPC error handling of gRPC API.

cc/ @yaron2 @msfussell @mchmarny

**Review by mchmarny — DISMISSED** [review]
/lgtm

File: pkg/grpc/api.go [hunk 3]
```diff
@@ -477,22 +476,22 @@ func (a *api) GetSecret(ctx context.Context, in *runtimev1pb.GetSecretRequest) (
 
 func (a *api) ExecuteStateTransaction(ctx context.Context, in *runtimev1pb.ExecuteStateTransactionRequest) (*empty.Empty, error) {
 	if a.stateStores == nil || len(a.stateStores) == 0 {
-		err := errors.New("ERR_STATE_STORE_NOT_CONFIGURED")
+		err := status.Error(codes.Unavailable, "state store is not configured")
```
Comment (mukundansundar) [inline pkg/grpc/api.go:479]: I think `FAILED_PRECONDITION` makes more sense than Unavailable.

**Comment by mukundansundar** [comment]
@beiwei30 Can you add unit test for the negative scenarios?

File: pkg/grpc/api.go (hunk 1 above)
Comment (beiwei30) [reply pkg/grpc/api.go]: done.

File: pkg/grpc/api.go (hunk 2 above)
Comment (beiwei30) [reply pkg/grpc/api.go]: done

File: pkg/grpc/api.go (hunk 3 above)
Comment (beiwei30) [reply pkg/grpc/api.go]: done.

**Comment by beiwei30** [comment]
> @beiwei30 Can you add unit test for the negative scenarios?

Sure, I've filed another issue for this task. Pls. assign it to me.

**Comment by beiwei30** [comment]
@youngbupark pls. check the proposed change again. It looks like this pull request has not been fully merged yet. Once it's merged, I can start to work on ##2268

**Review by yaron2 — APPROVED** [review]

**Review by youngbupark — APPROVED** [review]
Thanks for the contribution.
